# -------------------------------------------------------------------------------------------------
# Copyright (c) 2021, DHS. This file is part of YEAT: http://github.com/bioforensics/yeat
#
# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
# Development Center.
# -------------------------------------------------------------------------------------------------

from argparse import Action, ArgumentParser
import json
from pathlib import Path
from pkg_resources import resource_filename
from snakemake import snakemake
import sys
import yeat


class InitAction(Action):
    def __call__(self, parser, namespace, values, option_string=None):
        config = [
            dict(
                label="assembly1",
                algorithm="spades",
            ),
            # dict(
            #     label="assembly2",
            #     algorithm="megahit",
            # ),
        ]
        json.dump(config, sys.stdout, indent=4)
        raise SystemExit()


def get_parser():
    parser = ArgumentParser()
    parser.add_argument("-v", "--version", action="version", version=f"YEAT v{yeat.__version__}")
    parser.add_argument(
        "-O",
        "--outdir",
        type=str,
        metavar="DIR",
        default=".",
        help="output directory; default is '.'",
    )
    parser.add_argument(
        "-t",
        "--threads",
        type=int,
        metavar="T",
        default=1,
        help="number of threads for Snakemake; default is 1",
    )
    parser.add_argument(
        "--sample",
        type=str,
        metavar="S",
        default="sample",
        help="sample name for Snakemake; default is 'sample'",
    )
    parser.add_argument(
        "-n",
        "--dry-run",
        action="store_true",
        help="construct workflow DAG and print a summary but do not execute",
    )
    parser.add_argument(
        "config",
        type=str,
        help="genome assembly configuration in JSON format; template can be generated by running `yeat --init`",
    )
    parser.add_argument("reads", type=str, nargs=2, help="paired-end reads in FASTQ format")
    parser.add_argument(
        "--init",
        action=InitAction,
        nargs=0,
        help="print a template database config file to the terminal (stdout) and exit",
    )
    return parser


def run_spades(read1, read2, outdir=".", cores=1, sample="sample", dryrun="dry"):
    snakefile = resource_filename("yeat", "assembly/Snakefile_Spades")
    r1 = Path(read1).resolve()
    r2 = Path(read2).resolve()
    config = dict(
        read1=r1,
        read2=r2,
        outdir=outdir,
        cores=cores,
        sample=sample,
        dryrun=dryrun,
    )
    success = snakemake(
        snakefile,
        config=config,
        cores=cores,
        dryrun=dryrun,
        printshellcmds=True,
        workdir=outdir,
    )
    if not success:
        raise RuntimeError("Snakemake Failed")
